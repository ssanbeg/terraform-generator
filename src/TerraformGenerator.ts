import child_process from 'child_process';
import fs from 'fs';
import path from 'path';
import shell from 'shelljs';
import { Util } from './Util';
import { Block, Resource, Data, Module, Output, Provider, Variable, Backend, Provisioner, ResourceToDataOptions, Locals } from '.';

/**
 * @param dir directoty, default = .
 * @param tfFilename
 * @param tfvarsFilename
 * @param format
 */
export interface WriteOptions {
  /**
   * Directory to write to.
   * Default = .
   */
  dir?: string;
  /**
   * Terraform filename, must ends with .tf.
   * Default = terraform.tf
   */
  tfFilename?: string;
  /**
   * Terraform variables filename, must ends with .tfvars.
   * Default = terraform.tfvars
   */
  tfvarsFilename?: string;
  /**
   * Put true to use 'terraform fmt' to format the configuration, Terraform must be installed.
   * If the terraform binary is named differently in your machine, put the binary name instead.
   * Default = false
   */
  format?: boolean | string;
}

export class TerraformGenerator {

  private readonly arguments: Record<string, any>;
  private readonly blocks: Block[] = [];
  private variables: Record<string, any> = {};

  /**
   * Construct Terraform generator.
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  constructor(args?: Record<string, any>) {
    this.arguments = args;
  }

  /**
   * Generate Terraform configuration as string.
   */
  generate(): { tf: string; tfvars: string } {
    return {
      tf: this.generateTf(),
      tfvars: Object.keys(this.variables).length > 0 ? this.generateTfvars() : null
    };
  }

  private generateTf(): string {
    let str = '';

    if (this.arguments || this.blocks.filter(block => block instanceof Backend).length > 0) {
      str += 'terraform {\n';
      str += Util.argumentsToString(this.arguments);
      this.blocks.forEach(block => {
        if (block instanceof Backend) {
          str += block.toTerraform();
        }
      });
      str += '}\n\n';
    }

    this.blocks.forEach(block => {
      if (!(block instanceof Backend)) {
        str += block.toTerraform();
      }
    });

    return Util.unescape(str);
  }

  private generateTfvars(): string {
    let str = '';

    Object.keys(this.variables).forEach(key => {
      str += Util.argumentsToString({ [key]: this.variables[key] });
      str += '\n';
    });

    return str;
  }

  /**
   * Write Terraform configuration to a file.
   *
   * @param options options
   */
  write(options?: WriteOptions): void {
    if (!options) {
      options = {};
    }
    if (!options.dir) {
      options.dir = '.';
    }
    if (!options.tfFilename) {
      options.tfFilename = 'terraform.tf';
    }
    if (!options.tfFilename.endsWith('.tf')) {
      options.tfFilename += '.tf';
    }
    if (!options.tfvarsFilename) {
      options.tfvarsFilename = 'terraform.tfvars';
    }
    if (!options.tfvarsFilename.endsWith('.tfvars')) {
      options.tfvarsFilename += '.tfvars';
    }
    if (options.format === true) {
      options.format = 'terraform';
    }

    const result = this.generate();
    shell.mkdir('-p', options.dir);
    fs.writeFileSync(path.join(options.dir, options.tfFilename), result.tf);
    if (result.tfvars) {
      fs.writeFileSync(path.join(options.dir, options.tfvarsFilename), result.tfvars);
    }

    if (options.format) {
      child_process.execSync(`${options.format} fmt`, { cwd: options.dir });
    }
  }

  /**
   * Add blocks into Terraform.
   *
   * @param blocks blocks
   */
  addBlocks(...blocks: Block[]): this {
    blocks.forEach(block => this.blocks.push(block));
    return this;
  }

  /**
   * Add provider into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param args arguments
   */
  provider(type: string, args?: Record<string, any>): Provider {
    const block = new Provider(type, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add resource into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param name name
   * @param args arguments
   * @param provisioners provisioners
   */
  resource(type: string, name: string, args?: Record<string, any>, provisioners?: Provisioner[]): Resource {
    const block = new Resource(type, name, args, provisioners);
    this.addBlocks(block);
    return block;
  }

  /**
   * Convert resource into data source and add it into Terraform.
   *
   * @param resource resource
   * @param options options
   * @param argNames names of resource arguments to be converted into data source arguments;
   * use array for name mapping, position 0 = original resource's argument name, position 1 = mapped data source's argument name
   * @param args extra arguments
   */
  dataFromResource(resource: Resource, options: ResourceToDataOptions, argNames: (string | [string, string])[],
    args?: Record<string, any>): Data {
    const block = resource.toData(options, argNames, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add data source into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param name name
   * @param args arguments
   */
  data(type: string, name: string, args?: Record<string, any>): Data {
    const block = new Data(type, name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add module into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param name name
   * @param args arguments
   */
  module(name: string, args?: Record<string, any>): Module {
    const block = new Module(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add output into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param name name
   * @param args arguments
   */
  output(name: string, args?: Record<string, any>): Output {
    const block = new Output(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add locals into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  locals(args?: Record<string, any>): Locals {
    const block = new Locals(args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add variable into Terraform.
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param name name
   * @param args arguments
   * @param value variable value
   */
  variable(name: string, args?: Record<string, any>, value?: any): Variable {
    const block = new Variable(name, args);
    this.addBlocks(block);
    if (value != null) {
      this.addVars({ [name]: value });
    }
    return block;
  }

  /**
   * Add backend into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param args arguments
   */
  backend(type: string, args?: Record<string, any>): Backend {
    const block = new Backend(type, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add provisioner into Terraform.
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param args arguments
   */
  provisioner(type: string, args?: Record<string, any>): Provisioner {
    const block = new Provisioner(type, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add variable values into Terraform.
   *
   * @param variables variables
   */
  addVars(variables: Record<string, any>): this {
    this.variables = {
      ...this.variables,
      ...variables
    };
    return this;
  }

  /**
   * Merge this instance with other TerraformGenerator instances.
   *
   * @param tfgs other instances
   */
  merge(...tfgs: TerraformGenerator[]): this {
    tfgs.forEach(tfg => {
      this.addBlocks(...tfg.getBlocks());
      this.addVars(tfg.getVars());
    });
    return this;
  }

  /**
   * Get arguments.
   */
  getArguments(): Record<string, any> {
    return this.arguments;
  }

  /**
   * Get blocks.
   */
  getBlocks(): Block[] {
    return this.blocks;
  }

  /**
   * Get variables.
   */
  getVars(): Record<string, any> {
    return this.variables;
  }

}
