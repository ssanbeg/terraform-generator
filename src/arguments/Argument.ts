import { Util } from '../Util';

export class Argument {

  readonly argument: string | Argument;

  /**
   * Construct argument.
   *
   * @param arg argument as string or copy from another argument object
   */
  constructor(arg: string | Argument) {
    if (!arg || (typeof arg === 'string' && !arg.trim())) {
      throw new Error('Argument cannot be empty.');
    }

    this.argument = arg;
  }

  /**
   * Get argument's attribute.
   *
   * @param name attribute name
   */
  attr(name: string): Argument {
    name = name.trim();
    if (this.argument instanceof Argument) {
      return this.argument.attr(name);
    }
    return new Argument(`${this.argument}.${name}`);
  }

  /**
   * Get list argument's element.
   *
   * @param idx element index
   */
  element(idx: number): Argument {
    if (this.argument instanceof Argument) {
      return this.argument.element(idx);
    }
    return new Argument(`${this.argument}[${idx}]`);
  }

  /**
   * To Terraform representation.
   * Use this function when argument is used as an interpolation in another Terraform argument or code.
   */
  toTerraform(): string {
    let str = '';
    if (this.argument instanceof Argument) {
      str += this.argument.toTerraform();
    } else {
      str += this.argument;
    }
    return Util.escape(str);
  }

  /**
   * To string.
   * Use this function when argument is used as an interpolation in a Terraform string or heredoc.
   * It is automatically called when argument is used in template literal.
   */
  toString(): string {
    return `\${${this.toTerraform()}}`;
  }

}

/**
 * Convenient function to construct new argument.
 *
 * @param arg argument as string or copy from another argument object
 */
export const arg = (arg: string | Argument): Argument => new Argument(arg);
