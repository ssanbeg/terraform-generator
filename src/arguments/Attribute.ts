import { Argument, Block } from '..';

export class Attribute extends Argument {

  /**
   * Construct block's attribute.
   *
   * @param block block
   * @param attrName attribute name
   */
  constructor(block: Block, attrName: string) {
    super(Attribute.constructArgument(block, attrName));
  }

  private static constructArgument(block: Block, attrName: string): string {
    if (!block) {
      throw new Error('Attribute block cannot be null.');
    }
    if (!attrName || !attrName.trim()) {
      throw new Error('Attribute name cannot be empty.');
    }

    return `${block.asArgument().toTerraform()}.${attrName.trim()}`;
  }

}

/**
 * Convenient function to construct new block's attribute.
 *
 * @param block block
 * @param name attribute name
 */
export const attr = (block: Block, name: string): Attribute => new Attribute(block, name);
