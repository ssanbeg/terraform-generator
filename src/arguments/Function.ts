import { Argument } from '..';
import { Util } from '../Util';

export class Function extends Argument {

  /**
   * Construct function argument.
   *
   * @param name function name
   * @param args function arguments
   */
  constructor(name: string, ...args: any[]) {
    super(Function.constructArgument(name, ...args));
  }

  private static constructArgument(fn: string, ...args: any[]): string {
    if (!fn || !fn.trim()) {
      throw new Error('Function name cannot be empty.');
    }
    if (args.filter(arg => arg == null).length > 0) {
      throw new Error(`Invalid function argument: ${args}`);
    }

    let str = `${fn}(`;
    args.forEach((arg, i) => {
      str += Util.argumentValueToString(arg);
      if (i < args.length - 1) {
        str += ', ';
      }
    });
    str += ')';
    return str;
  }

}

/**
 * Convenient function to construct new function.
 *
 * @param name function name
 * @param args function arguments
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export const fn = (name: string, ...args: any[]): Function => new Function(name, ...args);
