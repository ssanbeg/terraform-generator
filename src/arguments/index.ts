export * from './Argument';
export * from './Attribute';
export * from './Function';
export * from './Heredoc';
