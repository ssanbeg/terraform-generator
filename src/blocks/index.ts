export * from './Block';
export * from './Backend';
export * from './Data';
export * from './Locals';
export * from './Module';
export * from './Output';
export * from './Provider';
export * from './Provisioner';
export * from './Resource';
export * from './Variable';
