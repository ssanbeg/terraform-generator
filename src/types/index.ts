export * from './List';
export * from './Map';
