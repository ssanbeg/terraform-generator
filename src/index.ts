import { TerraformGenerator } from './TerraformGenerator';
export default TerraformGenerator;

export { WriteOptions } from './TerraformGenerator';

export { Block } from './blocks/Block';
export { Provider } from './blocks/Provider';
export { Resource, ResourceToDataOptions } from './blocks/Resource';
export { Data } from './blocks/Data';
export { Output } from './blocks/Output';
export { Module } from './blocks/Module';
export { Locals } from './blocks/Locals';
export { Variable } from './blocks/Variable';
export { Backend } from './blocks/Backend';
export { Provisioner } from './blocks/Provisioner';

export { Argument, arg } from './arguments/Argument';
export { Attribute, attr } from './arguments/Attribute';
export { Function, fn } from './arguments/Function';
export { Heredoc, heredoc } from './arguments/Heredoc';

export { Map, map } from './types/Map';
export { List, list } from './types/List';
