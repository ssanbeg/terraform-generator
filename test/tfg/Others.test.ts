/* eslint-disable @typescript-eslint/naming-convention */
import fs from 'fs';
import path from 'path';
import { arg } from '../../src/arguments';
import { Provisioner } from '../../src/blocks';
import { TerraformGenerator } from '../../src/TerraformGenerator';
import { map } from '../../src/types';

const createTerraformGenerator = (): TerraformGenerator => {
  const tfg = new TerraformGenerator();

  tfg.variable('test', {
    type: arg('string')
  });

  tfg.variable('test2', {
    type: arg('string')
  }, 'test');

  tfg.data('aws_vpc', 'test', {
    cidr_block: 'test'
  });

  tfg.module('test', {
    source: './test'
  });

  const r = tfg.resource('aws_vpc', 'test', {
    cidr_block: 'test',
    tags: map({
      a: 'a'
    })
  });

  tfg.dataFromResource(r, undefined, ['cidr_block', ['tags', 'tag']]);
  tfg.dataFromResource(r, { name: 'test2' }, ['cidr_block', ['tags', 'tag']]);

  tfg.resource('innerBlock', 'innerBlock', {
    a: 'a'
  }, [
    new Provisioner('local-exec', {
      command: 'echo hello'
    }),
    new Provisioner('local-exec', {
      command: 'echo world'
    })
  ]);

  const locals = tfg.locals({
    a: 'a',
    b: 123,
    c: r.attr('x')
  });
  tfg.resource('locals', 'locals', {
    a: locals.arg('a')
  });

  tfg.provisioner('provisioner', {
    a: 'a',
    b: 123,
    c: r.attr('x')
  });

  tfg.resource('tags', 'tags', {
    tags: map({
      'a': 'a',
      'b c d': 'b c d'
    })
  });

  const tfg2 = new TerraformGenerator();
  tfg2.resource('tfg2', 'tfg2', {
    tfg2: 'tfg2'
  });

  tfg.merge(tfg2);

  return tfg;
};

const outputDir = path.join('test', '__output__');

test('Others', () => {
  const tfg = createTerraformGenerator();

  expect(tfg.getArguments()).toMatchSnapshot();
  expect(tfg.getBlocks()).toMatchSnapshot();
  expect(tfg.getVars()).toMatchSnapshot();
  expect(tfg.generate()).toMatchSnapshot();

  tfg.write({ dir: outputDir });
  const tf = fs.readFileSync(path.join(outputDir, 'terraform.tf'), 'utf8');
  expect(tf).toMatchSnapshot();
  const tfvars = fs.readFileSync(path.join(outputDir, 'terraform.tfvars'), 'utf8');
  expect(tfvars).toMatchSnapshot();
});

afterAll(() => {
  fs.unlinkSync(path.join(outputDir, 'terraform.tf'));
  fs.unlinkSync(path.join(outputDir, 'terraform.tfvars'));
  fs.rmdirSync(outputDir);
});
