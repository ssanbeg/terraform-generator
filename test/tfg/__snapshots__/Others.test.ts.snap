// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Others 1`] = `undefined`;

exports[`Others 2`] = `
Array [
  Variable {
    "arguments": Object {
      "type": Argument {
        "argument": "string",
      },
    },
    "blockNames": Array [
      "test",
    ],
    "blockType": "variable",
    "innerBlocks": Array [],
    "name": "test",
  },
  Variable {
    "arguments": Object {
      "type": Argument {
        "argument": "string",
      },
    },
    "blockNames": Array [
      "test2",
    ],
    "blockType": "variable",
    "innerBlocks": Array [],
    "name": "test2",
  },
  Data {
    "arguments": Object {
      "cidr_block": "test",
    },
    "blockNames": Array [
      "aws_vpc",
      "test",
    ],
    "blockType": "data",
    "innerBlocks": Array [],
    "name": "test",
    "type": "aws_vpc",
  },
  Module {
    "arguments": Object {
      "source": "./test",
    },
    "blockNames": Array [
      "test",
    ],
    "blockType": "module",
    "innerBlocks": Array [],
    "name": "test",
  },
  Resource {
    "arguments": Object {
      "cidr_block": "test",
      "tags": Map {
        "arguments": Object {
          "a": "a",
        },
      },
    },
    "blockNames": Array [
      "aws_vpc",
      "test",
    ],
    "blockType": "resource",
    "innerBlocks": Array [],
    "name": "test",
    "type": "aws_vpc",
  },
  Data {
    "arguments": Object {
      "cidr_block": "test",
      "filter": Array [
        Object {
          "name": "tag:a",
          "values": Array [
            "a",
          ],
        },
      ],
    },
    "blockNames": Array [
      "aws_vpc",
      "test",
    ],
    "blockType": "data",
    "innerBlocks": Array [],
    "name": "test",
    "type": "aws_vpc",
  },
  Data {
    "arguments": Object {
      "cidr_block": "test",
      "filter": Array [
        Object {
          "name": "tag:a",
          "values": Array [
            "a",
          ],
        },
      ],
    },
    "blockNames": Array [
      "aws_vpc",
      "test2",
    ],
    "blockType": "data",
    "innerBlocks": Array [],
    "name": "test2",
    "type": "aws_vpc",
  },
  Resource {
    "arguments": Object {
      "a": "a",
    },
    "blockNames": Array [
      "innerBlock",
      "innerBlock",
    ],
    "blockType": "resource",
    "innerBlocks": Array [
      Provisioner {
        "arguments": Object {
          "command": "echo hello",
        },
        "blockNames": Array [
          "local-exec",
        ],
        "blockType": "provisioner",
        "innerBlocks": Array [],
        "type": "local-exec",
      },
      Provisioner {
        "arguments": Object {
          "command": "echo world",
        },
        "blockNames": Array [
          "local-exec",
        ],
        "blockType": "provisioner",
        "innerBlocks": Array [],
        "type": "local-exec",
      },
    ],
    "name": "innerBlock",
    "type": "innerBlock",
  },
  Locals {
    "arguments": Object {
      "a": "a",
      "b": 123,
      "c": Attribute {
        "argument": "aws_vpc.test.x",
      },
    },
    "blockNames": Array [],
    "blockType": "locals",
    "innerBlocks": Array [],
  },
  Resource {
    "arguments": Object {
      "a": Argument {
        "argument": "local.a",
      },
    },
    "blockNames": Array [
      "locals",
      "locals",
    ],
    "blockType": "resource",
    "innerBlocks": Array [],
    "name": "locals",
    "type": "locals",
  },
  Provisioner {
    "arguments": Object {
      "a": "a",
      "b": 123,
      "c": Attribute {
        "argument": "aws_vpc.test.x",
      },
    },
    "blockNames": Array [
      "provisioner",
    ],
    "blockType": "provisioner",
    "innerBlocks": Array [],
    "type": "provisioner",
  },
  Resource {
    "arguments": Object {
      "tags": Map {
        "arguments": Object {
          "a": "a",
          "b c d": "b c d",
        },
      },
    },
    "blockNames": Array [
      "tags",
      "tags",
    ],
    "blockType": "resource",
    "innerBlocks": Array [],
    "name": "tags",
    "type": "tags",
  },
  Resource {
    "arguments": Object {
      "tfg2": "tfg2",
    },
    "blockNames": Array [
      "tfg2",
      "tfg2",
    ],
    "blockType": "resource",
    "innerBlocks": Array [],
    "name": "tfg2",
    "type": "tfg2",
  },
]
`;

exports[`Others 3`] = `
Object {
  "test2": "test",
}
`;

exports[`Others 4`] = `
Object {
  "tf": "variable \\"test\\"{
type = string
}

variable \\"test2\\"{
type = string
}

data \\"aws_vpc\\" \\"test\\"{
cidr_block = \\"test\\"
}

module \\"test\\"{
source = \\"./test\\"
}

resource \\"aws_vpc\\" \\"test\\"{
cidr_block = \\"test\\"
tags = {
a = \\"a\\"
}
}

data \\"aws_vpc\\" \\"test\\"{
filter {
name = \\"tag:a\\"
values = [
\\"a\\"
]
}
cidr_block = \\"test\\"
}

data \\"aws_vpc\\" \\"test2\\"{
filter {
name = \\"tag:a\\"
values = [
\\"a\\"
]
}
cidr_block = \\"test\\"
}

resource \\"innerBlock\\" \\"innerBlock\\"{
a = \\"a\\"
provisioner \\"local-exec\\"{
command = \\"echo hello\\"
}
provisioner \\"local-exec\\"{
command = \\"echo world\\"
}
}

locals{
a = \\"a\\"
b = 123
c = aws_vpc.test.x
}

resource \\"locals\\" \\"locals\\"{
a = local.a
}

provisioner \\"provisioner\\"{
a = \\"a\\"
b = 123
c = aws_vpc.test.x
}

resource \\"tags\\" \\"tags\\"{
tags = {
a = \\"a\\"
\\"b c d\\" = \\"b c d\\"
}
}

resource \\"tfg2\\" \\"tfg2\\"{
tfg2 = \\"tfg2\\"
}

",
  "tfvars": "test2 = \\"test\\"

",
}
`;

exports[`Others 5`] = `
"variable \\"test\\"{
type = string
}

variable \\"test2\\"{
type = string
}

data \\"aws_vpc\\" \\"test\\"{
cidr_block = \\"test\\"
}

module \\"test\\"{
source = \\"./test\\"
}

resource \\"aws_vpc\\" \\"test\\"{
cidr_block = \\"test\\"
tags = {
a = \\"a\\"
}
}

data \\"aws_vpc\\" \\"test\\"{
filter {
name = \\"tag:a\\"
values = [
\\"a\\"
]
}
cidr_block = \\"test\\"
}

data \\"aws_vpc\\" \\"test2\\"{
filter {
name = \\"tag:a\\"
values = [
\\"a\\"
]
}
cidr_block = \\"test\\"
}

resource \\"innerBlock\\" \\"innerBlock\\"{
a = \\"a\\"
provisioner \\"local-exec\\"{
command = \\"echo hello\\"
}
provisioner \\"local-exec\\"{
command = \\"echo world\\"
}
}

locals{
a = \\"a\\"
b = 123
c = aws_vpc.test.x
}

resource \\"locals\\" \\"locals\\"{
a = local.a
}

provisioner \\"provisioner\\"{
a = \\"a\\"
b = 123
c = aws_vpc.test.x
}

resource \\"tags\\" \\"tags\\"{
tags = {
a = \\"a\\"
\\"b c d\\" = \\"b c d\\"
}
}

resource \\"tfg2\\" \\"tfg2\\"{
tfg2 = \\"tfg2\\"
}

"
`;

exports[`Others 6`] = `
"test2 = \\"test\\"

"
`;
